python/ray/ml/utils/remote_storage.py

```diff
@@ -0,0 +1,142 @@
+import re
+from typing import Optional, Tuple
+
+try:
+    import fsspec
+except ImportError:
+    fsspec = None
+
+try:
+    import pyarrow
+    import pyarrow.fs
+except (ImportError, ModuleNotFoundError):
+    pyarrow = None
+
+from ray import logger
+
+URI_PROTOCOL_REGEX = r"^(\w+)://(.+)$"
+S3_PREFIX = "s3://"
+GS_PREFIX = "gs://"
+HDFS_PREFIX = "hdfs://"
+ALLOWED_REMOTE_PREFIXES = (S3_PREFIX, GS_PREFIX, HDFS_PREFIX)
+
+
+def _assert_pyarrow_installed():
+    if pyarrow is None:
+        raise RuntimeError(
+            "Uploading, downloading, and deleting from cloud storage "
+            "requires pyarrow to be installed. Install with: "
+            "`pip install pyarrow`. Subsequent calls to cloud operations "
+            "will be ignored."
+        )
+
+
+def _split_protocol_path(uri: str) -> Tuple[Optional[str], Optional[str]]:
+    protocol_match = re.match(URI_PROTOCOL_REGEX, uri)
+    if not protocol_match:
+        return None, None
+
+    return protocol_match.group(1), protocol_match.group(2)
+
+
+def fs_hint(uri: str) -> str:
+    """Return a hint how to install required filesystem package"""
+    if pyarrow is None:
+        return "Please make sure PyArrow is installed: `pip install pyarrow`."
+    if fsspec is None:
+        return "Try installing fsspec: `pip install fsspec`."
+
+    from fsspec.registry import known_implementations
+
+    protocol, _path = _split_protocol_path(uri)
+
+    if protocol in known_implementations:
+        return known_implementations[protocol]["err"]
+
+    return "Make sure to install and register your fsspec-compatible filesystem."
```

Comment:
It means the specified protocol is not registered, which is usually the case when it's not installed. But because we don't know anything about the protocol (as it's not in known_implementations), we can only instruct the user to install whatever it is he wants to use